fix(logic): Check for negated relationals after flattening

oscarbenjamin opened the pull request:
Previously And(a, b, c, d) would check for negated Relationals like
And(x>1,x<=1) and would flatten nested Ands like And(And(a, b), And(c,
d)). However flattening would happen after checking for negated
Relationals so would miss the negated Relationals in e.g.
And(And(x<1, y>2), And(x>=1, z>2)). This commit ensures that checking
for negated Relationals happens after flattening nested Ands.

On master we have:
```julia
In [2]: a, b = symbols('a, b')                                                                                                                                

In [3]: (Eq(a, 1) & Ne(b, 1)) & (Ne(a, 1) & Eq(b, 2))                                                                                                         
Out[3]: a = 1 ∧ b = 2 ∧ a ≠ 1 ∧ b ≠ 1
```
This PR gives
```julia
In [1]: a, b = symbols('a, b')                                                                                                                                

In [2]: (Eq(a, 1) & Ne(b, 1)) & (Ne(a, 1) & Eq(b, 2))                                                                                                         
Out[2]: False
```

<!-- BEGIN RELEASE NOTES -->
* logic
    * Opposing Relationals are now detected correctly in nested Ands allowing evaluation to False in more cases.
<!-- END RELEASE NOTES -->

oscarbenjamin committed:
fix(logic): Check for negated relationals after flattening

Previously And(a, b, c, d) would check for negated Relationals like
And(x>1,x<=1) and would flatten nested Ands like And(And(a, b), And(c,
d)). However flattening would happen after checking for negated
Relationals so would miss the negated Relationals in e.g.
And(And(x<1, y>2), And(x>=1, z>2)). This commit ensures that checking
for negated Relationals happens after flattening nested Ands.

smichr commented on sympy/logic/boolalg.py:
```diff
@@ -679,20 +679,21 @@ class And(LatticeOp, BooleanFunction):
+        args = LatticeOp._new_args_filter(args, And)
+        newargs = []
+        rel = set()
+        for x in args:
```
```suggestion
        for x in ordered(arg)s:
```

Since we are going to be filtering args, let's make it canonical:
```python
>>> set([And(*i) for i in permutations((l,g,le,ge))])
{(A >= B) & (A > B)}
```
(`ordered` is already imported so this diff can be added without additional edits; the test of permutations should be added, though)

oscarbenjamin replied:
I'm a bit confused by this bit of code. Previously there was `reversed` rather than `ordered`. Removing reversed caused the only test failure that I saw here. Do you remember why the `reversed` was added?

asmeurer replied:
Unfortunately there was a syntaxerror with this change.

smichr replied:
I think the `reversed` may have been an attempt to make something matching a canonical form be the first form that was seen. But I think canonically predictable is better than predictable.

oscarbenjamin committed:
fix(logic): Canonicalise Relational filtering in And

Using ordered here ensures that when redundant Relationals are filtered
from an And we always end up with the same canonicalised args:

  >>> set([And(*i) for i in permutations((l,g,le,ge))])
  {(A >= B) & (A > B)}

Co-Authored-By: Christopher Smith <smichr@gmail.com>
